### PR TITLE
Towards full metadata - fully declare 'bespoke' tokens

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Towards full metadata - fully declare 'bespoke' tokens

The goal is to declare all tokens with full metadata and cache that - rather than the convoluted ways we have of looking up metadata  - I'm working towards https://github.com/civicrm/civicrm-core/pull/21706 but working through roadblocks

Before
----------------------------------------

'bespokeTokens' (only in master) declared as a flat array

After
----------------------------------------
declared as a nuanced array

Technical Details
----------------------------------------
This functionality all has really good test cover. One thing it does change is that membership & event tokens start to advertise the 
machine name tokens - the intent is to get to the point where ALL fields are declared as token metadata (which will be cached) but only those with 'audience' = 'user' will be in the widget - see https://github.com/civicrm/civicrm-core/pull/21706 

Comments
----------------------------------------
@demeritcowboy @colemanw if you have a better function name that 'getBespokeFunctions` please speak now

variants on `custom` are too close to custom fields & calculated is not necessarily accurate.

### CODE REVIEW

**Comment by colemanw** [comment]
So I wonder about calculating these fields at the APIv4 level rather than in the token layer. For example, #21718 demonstrates how easy it is to add a calculated field to APIv4 which acts very much like a "real" field (that is, you can use it in any clause - select, where, having, order by).
We could do something similar for things like `Participant.balance` (or `amount_owed` or whatever)

**Comment by eileenmcnaughton** [comment]
@colemanw yeah - but not this round

**Comment by eileenmcnaughton** [comment]
@colemanw longer answer - I have some real doubts about the validity of that participant balance token - but  would would be REALLY useful is a similar token on the contribution side for balance - I had thought to add that if I can get far enough before the rc is cut. But for now I'm treating those tokens as ones we are stuck with

File: CRM/Member/Tokens.php [hunk 1]
```diff
@@ -76,4 +76,24 @@ public function getDependencies(): array {
...
+   * Get any tokens with custom calculation.
+   *
+   * In this case 'fee' should be converted to{membership.membership_type_id.fee}
+   * but we don't have the formatting support to do that with no
+   * custom intervention yet.
+   */
+  protected function getBespokeTokens(): array {
+    return [
+      'fee' => [
+        'title' => ts('Membership Fee'),
```
Comment (eileenmcnaughton) [inline CRM/Member/Tokens.php:89]: If we get to the point where the number formatting is settled I would deprecated this in favour of `{membership.membership_type_id.minimum_fee}`

File: CRM/Activity/Tokens.php [hunk 2]
```diff
@@ -162,6 +148,29 @@ public function getBasicTokens(): array {
...
+   * Get tokens that are special or calculated for this enitty.
+   *
+   * @return array|array[]
+   */
+  protected function getBespokeTokens(): array {
+    $tokens = [];
+    if (array_key_exists('CiviCase', CRM_Core_Component::getEnabledComponents())) {
+      $tokens['case_id'] = ts('Activity Case ID');
+      return [
+        'case_id' => [
```
Comment (eileenmcnaughton) [inline CRM/Activity/Tokens.php:161]: This token exists because it was added in the past but there are 1-1 entity validilty issues

Comment (colemanw) [reply CRM/Activity/Tokens.php:161]: Right, at best it could get you *a* case id (but no guarantee which one)

**Comment by colemanw** [comment]
Ok let's merge this since it's
1. Tested
2. Not hindering a full APIv4 approach
3. Sounds like the right direction

**Comment by eileenmcnaughton** [comment]
thanks @colemanw

**Pull request merged**